llama-server : implement universal assisted decoding

g2mt opened the pull request:
This pull request implements [universal assisted decoding](https://huggingface.co/blog/universal_assisted_generation) in llama-server. This is a method for performing speculative decoding given a draft model whose tokenizer is incompatible with the main model, by decoding and reencoding the generated text between the two models.

It currently works, but some improvements can be made.

- Token healing can be done to fix any weirdness that may occur if the draft model generates tokens that doesn't lie on a word boundary (not sure how much this affects performance).
- The translation process can be cached to improve sampling time, however it might require substantial refactoring.

g2mt committed:
llama-server : implement universal assisted decoding

g2mt committed:
Merge branch 'master' into master

jukofyork commented:
This looks really interesting! It's surprising how much crossover there is between many models' tokenisers.

g2mt committed:
Merge remote-tracking branch 'fork/master' into universal-decoding

g2mt committed:
Erase prompt tail for kv-cache

g2mt committed:
set vocab_dft_compatible in common_speculative

g2mt committed:
rename ctx_main to ctx_tgt

g2mt committed:
move vocab_dft_compatible to spec struct

g2mt committed:
clear mem_dft, remove mem

g2mt committed:
detokenize id_last for incompatible models

g2mt committed:
update comment

g2mt committed:
add --spec-replace flag

g2mt committed:
accept special tokens when translating between draft/main models

g2mt committed:
Merge remote-tracking branch 'upstream/master'

Mushoz commented:
Why was this closed? Was really looking forward to this

g2mt commented:
Anyways, I merged changes from upstream and updated the PR. The assigned labels are wrong and I don't know how to fix them.

Here are some sample outputs:

<details>
<summary>

```
./build/bin/llama-speculative-simple  -md ./pythia-160m.Q8_0.gguf  -m  ./Qwen2.5-Coder-0.5B-Instruct-Q8_0.gguf -p "The quick brown fox jumped over the lazy dog.\nThe quick brown fox jumped over the lazy dog.\nThe"
```

</summary>

```

The quick brown fox jumped over the lazy dog.
The quick brown fox jumped over the lazy dog.
The quick brown fox jumped over the lazy dog.
The quick brown fox jumped over the

encoded   21 tokens in    0.081 seconds, speed:  260.130 t/s
decoded   16 tokens in    0.318 seconds, speed:   50.335 t/s

n_draft   = 16
n_predict = 16
n_drafted = 15
n_accept  = 15
accept    = 100.000%

draft:

llama_perf_context_print:        load time =     172.40 ms
llama_perf_context_print: prompt eval time =      30.86 ms /    23 tokens (    1.34 ms per token,   745.20 tokens per second)
llama_perf_context_print:        eval time =     206.02 ms /    15 runs   (   13.73 ms per token,    72.81 tokens per second)
llama_perf_context_print:       total time =     399.07 ms /    38 tokens

target:

llama_perf_sampler_print:    sampling time =       0.97 ms /    16 runs   (    0.06 ms per token, 16511.87 tokens per second)
llama_perf_context_print:        load time =     475.29 ms
llama_perf_context_print: prompt eval time =     394.66 ms /    36 tokens (   10.96 ms per token,    91.22 tokens per second)
llama_perf_context_print:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:       total time =     571.55 ms /    37 tokens
```

</details>

I also added `--spec-replace` flag to replace the raw chat template tags between the main and the draft models. It only does a simple string replacement for each invocation, so it doesn't distinguish between special tokens and normal tokens.

<details>
<summary>

```
./build/bin/llama-speculative-simple -m gemma-3-1b-it-Q4_0.gguf -md  Llama-3.2-1B-Instruct-Q4_0.gguf -p "<start_of_turn>user\nRepeat the following sentence, as is:\nThe quick brown fox jumped over the lazy dog.<end_of_turn>\n<start_of_turn>model\nThe" -n 10 --spec-replace "<bos><start_of_turn>user\n" "<|eot_id|><|start_header_id|>user<|end_header_id|>\n\n" --spec-replace "<end_of_turn>\n<start_of_turn>model\n" "<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n"
```

</summary>

```
<bos><start_of_turn>user
Repeat the following sentence, as is:
The quick brown fox jumped over the lazy dog.<end_of_turn>
<start_of_turn>model
The quick brown fox jumped over the lazy dog.


encoded   29 tokens in    0.110 seconds, speed:  264.042 t/s
decoded   11 tokens in    0.850 seconds, speed:   12.947 t/s

n_draft   = 16
n_predict = 11
n_drafted = 26
n_accept  = 9
accept    = 34.615%

draft:

llama_perf_context_print:        load time =    2210.29 ms
llama_perf_context_print: prompt eval time =     113.77 ms /    29 tokens (    3.92 ms per token,   254.89 tokens per second)
llama_perf_context_print:        eval time =     572.07 ms /    25 runs   (   22.88 ms per token,    43.70 tokens per second)
llama_perf_context_print:       total time =     960.16 ms /    54 tokens

target:

llama_perf_sampler_print:    sampling time =       0.83 ms /    11 runs   (    0.08 ms per token, 13269.00 tokens per second)
llama_perf_context_print:        load time =     903.03 ms
llama_perf_context_print: prompt eval time =     724.68 ms /    56 tokens (   12.94 ms per token,    77.28 tokens per second)
llama_perf_context_print:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:       total time =    3170.51 ms /    57 tokens
```

</details>

g2mt committed:
Escape spec-replace

g2mt committed:
Merge branch 'ggml-org:master' into master

g2mt committed:
clamp draft result to size to params.n_draft

g2mt committed:
Merge branch 'ggml-org:master' into master

g2mt committed:
Merge branch 'ggml-org:master' into master

g2mt committed:
Merge branch 'ggml-org:master' into master

CISC commented on common/speculative.cpp:
```diff
@@ -134,32 +146,95 @@ bool common_speculative_are_compatible(
+
+        const llama_vocab * vocab_tgt = llama_model_get_vocab(model_tgt);
+        int32_t n_chars;
+        n_chars = llama_detokenize(vocab_tgt, &id_last, 1, &text[0], text.size(), false, true);
```
Odd way of saying you should use text.data(), but ok. :)

g2mt replied:
This section was based on the `common_detokenize` function:
https://github.com/ggml-org/llama.cpp/blob/master/common/common.cpp#L1350-L1360

I'm not sure if there's a better way of writing it.

g2mt committed:
fix comment

g2mt committed:
clean up code

CISC commented on tools/server/server.cpp:
```diff
@@ -1989,10 +1990,9 @@ struct server_context {
-                return false;
+            vocab_dft_compatible = common_speculative_are_compatible(ctx, llama_init_dft.context.get());
+            if (!vocab_dft_compatible) {
+                SRV_INF("the draft model '%s' is not compatible with the target model '%s'. tokens will be translated between the draft and target models.\n", params_base.speculative.model.path.c_str(), params_base.model.path.c_str());
```
Please add this to the examples as well.

g2mt committed:
restore old example

g2mt committed:
log common_speculative_are_compatible in speculative example

g2mt committed:
fix

g2mt committed:
Update common/speculative.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

g2mt committed:
Update common/speculative.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

g2mt committed:
Update common/speculative.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

g2mt commented on common/arg.cpp:
```diff
@@ -3249,6 +3253,13 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
             params.speculative.model.path = value;
         }
     ).set_examples({LLAMA_EXAMPLE_SPECULATIVE, LLAMA_EXAMPLE_SERVER}).set_env("LLAMA_ARG_MODEL_DRAFT"));
+    add_opt(common_arg(
+        {"--spec-replace"}, "TARGET", "DRAFT",
+        "translate the string in TARGET into DRAFT if the draft model and main model are not compatible",
+        [](common_params & params, const std::string & tgt, const std::string & dft) {
+            params.speculative.replacements.push_back({ tgt, dft });
+        }
+    ).set_examples({LLAMA_EXAMPLE_SPECULATIVE, LLAMA_EXAMPLE_SERVER}));
```
I added this in case translation improves speculation accuracy.

ggerganov commented:
Just make sure you ran some tests with vanilla speculative decoding and validate it is not broken.

CISC commented:
It worked earlier, though that was a few force-pushes ago, so hard to tell if anything broke, will test again.

CISC commented:
Hmmm, though server seems to be much slower with draft model?

g2mt commented:
I haven't changed anything related to the server between the force pushes. Just some fixes to the examples/speculative file.
